dask/dataframe/core.py

```diff
@@ -3782,8 +3782,8 @@ def apply_and_enforce(*args, **kwargs):
 
         if is_dataframe_like(df):
             # Need nan_to_num otherwise nan comparison gives False
-            if not np.array_equal(np.nan_to_num(meta.columns),
-                                  np.nan_to_num(df.columns)):
+            if not np.array_equal(np.sort(np.nan_to_num(meta.columns)),
+                                  np.sort(np.nan_to_num(df.columns))):
```

Comment:
What if the values aren't comparable? A mix of ints and strs, for example.

Reply:
Yeah right! I started thinking about how we can do this and found out: Shouldn't we be checking the dtypes of each column with the column name? But here we are checking only the column names. But instead, we can use:
```
meta.dtypes.to_dict() == df.dtypes.to_dict()
```
which works expectedly. Since we are comparing to dicts, order doesn't matter here and comparison always to true, when the content is same. 

Using that with np.nan as column names also work "magically" without converting to num, but I don't know how.
```
df = pd.DataFrame({"a": [1,2,3,4], np.nan: [10,20,30,40]})
ddf = pd.DataFrame({np.nan: [10,20,30,40], "a": [1,2,3,4]})
df.dtypes.to_dict() == df.dtypes.to_dict()  #True
```
So use of np.nan_to_num is also eliminated here!